krun/util.py

```diff
@@ -192,16 +207,28 @@ def read_popen_output_carefully(process, platform, print_stderr=True):
                 if stderr_consumer is not None:
                     stderr_consumer.send(d)
 
+        if timeout_at is not None and datetime.now() >= timeout_at:
+            raise ExecutionTimeout()
+
     # We know stderr and stdout are closed.
     # Now we are just waiting for the process to exit, which may have
     # already happened of course.
+    while True:
+        process.poll()
```

Comment:
Why not `wait` (or its equivalent) if stdout/stderr are closed at this point? Or does `poll` call `wait` underneath?

Reply:
It's a silly function name. [`poll()`](https://docs.python.org/2/library/subprocess.html#subprocess.Popen.poll) checks if the process has exited and sets the returncode if so. It does not block and is distinct from `poll(2)`.

There's also a [`wait()`](https://docs.python.org/2/library/subprocess.html#subprocess.Popen.wait), but it blocks forever.

In Python 3 (and `subprocess32`) there's [`wait()`](https://docs.python.org/3.5/library/subprocess.html#subprocess.Popen.wait) which has an optional timeout.

In the description (the EDIT part) I'm suggesting we perhaps use `subprocess32.wait()` with a timeout. I only found that half way through typing the PR description :)